### DISCUSSION TITLE
What I can and cannot do with Zimodem

### DISCUSSION THREAD

**Discussion by Biker1bob** [bozimmerman/Zimodem#99] [Q&A] [answered]
So, I have been using Zimodem for a few months now.

As a basic wifi connect - call out modem device it works fantastic. 

I cannot ..

Use it with any software so far to work as a host - this is outlined in the other thread I have. 

IF YOU are able to do either of the above items successfully I would like to hear from you!

**Comment by frnno967** [comment, marked as answer]
Transfers have been a persistent problem, and one which I believe is related to some bugs in the lower level code that Zimodem relies on. Jim Drew who makes the (non-open source) WiModem writes his own firmware which is similar to Zimodem in concept but implements much more functionality. He's found that some of the libraries that handle TCP communication have some subtle bugs that cause characters to drop at high throughput rates, and the problem isn't caused by flow control or lack thereof. In other words the hardware can handle the serial timing requirements just fine but it's in the handoff and packetization of the data inside the ESP that things are getting dropped. Bo has done an amazing job building this open source firmware but he hasn't worked to fix the problems in the lower level libraries that have contributed to data loss at high throughput rates, as is the case during transfers.

**Reply by Biker1bob** [reply]
Yes I own 2 of JDs modems a well - I like them very much its just that I built this one into an old Avatex 1200hc case. I also own 3 8266 node mcu - so could build a couple more for friends. Thanks for your response. I did manage to get transfer working.. but you are correct there were a number of CRC errors in the transfers that slowed things down.. and sometimes on ANSI screens the odd character here and there does get dropped.

**Reply by markaudacity** [reply]
I own a WiModem232, and it still drops characters at any rate faster than 300-600 baud (the issue does not seem to be consistent).
Mine also crashes and dumps core occasionally, which leads me to wonder if mine has a defective 8266 on it.

**Comment by Biker1bob** [comment]
I wish there was a bigger audience for Bo on this.. or that I myself knew how to fix the issues.

I cannot do any transfers at 19200 - 9600 maybe, slower than that and its really not worth it. 

I have contacted Jim.. see what I can work out there. I just want my DIY Avatex to be able to host more than anything.. and for whatever reason I cannot get ZImodem to work with any of the BBS software I am running on the Atari STs.

**Reply by frnno967** [reply]
Agreed! I'm trying to convince Jim to attend VCF Southwest and get him together with Bo so that maybe they could exchange notes on how to fix the underlying problem in Zimodem. Jim's stuff is great but we also really need an open source implementation that works 100% for those of us who sometimes just need to roll an ESP32 breadboard wifi modem for testing.. or more.

**Reply by Biker1bob** [reply]
Ill post some pics of what I have built. Jim says he is coming out with a new WImodem232 pro soon.  This has much more involved. Not really the way I want to go. He says he can send me a programmed modem with no RS232 on it ( the Avatex has one)  I could wire up the lines myself - I guess running them to the port and my Gate with resistors to the LEDs.  I want this to work as a bbs modem for that Zimodem is not close enough yet. Lets hope Jim wants to share some knowledge.. he has done a lot of work though, so I understand him wanting to keep it commercial.

**Comment by deltecent** [comment]
This is an old topic. When the computer sends data at 9600 baud to the Zimodem, the Zimodem cannot keep up and data eventually gets lost. I believe this is caused by `ZStream::socketWrite(uint8_t c)` being very slow. I fixed the problem by buffering up to 10 bytes in `ZStream::serialIncoming()` and using `ZStream::socketWrite(uint8_t *buf, uint8_t len)`. It now keeps up just fine.
